### ISSUE TITLE
Protobuf decoding error with simultaneous read/writes

### ISSUE THREAD

**Issue by litch** [mathieuravaux/eventstore-ruby#3]
I'm doing some exploration with this and it seems that I'm getting some behavior where if I'm writing to a Stream while a subscriber is reading from it, the subscriber is pretty fail-y.  If I pre-seed the stream, there doesn't seem to be a problem, but simultaneous read/writes seem to be problematic.

Is this something you have seen as well?

```
Protobuf decoding error
#<NoMethodError: undefined method `bytesize' for nil:NilClass>
{:type=>Eventstore::StreamEventAppeared, :message=>"\n)test-9676b89c-1b99-4854-a1c4-8fa42331873b\x10\a\x1A\x10\xC28\xD42v\xA0B\e\xBE 0\xB0|\xC9\x06\xCE\"\tTestEvent(\x010\x01:\x1D{\"at\":1425144059,\"foo\":\"bar\"}B\x00H\xAA\xDF\x99\xF1\xA0\xB2\x88\xE9\bP\xFC\xAF\xD2\x89\xBD)\x18\xC0\x9F\xCB\x01 \xC0\x9F\xCB\x01\xA2\x00\x00\x00\xC2\x00"}


/Users/litch/.rbenv/versions/jruby-1.7.15/lib/ruby/gems/shared/gems/beefcake-1.1.0.pre1/lib/beefcake/buffer/base.rb:86:in `length'
```

**Comment by mathieuravaux** [comment]
Hi Justin,

I've seen that as well lately, as well as disconnections problem while writing continuously to the EventStore. I think those happen when ES sends a HeartBeatRequest, and we somehow don't answer it in this case in a way the server expect.

I'll try a solution to both problems today or tomorrow; in a gist, all the ES read/writes would be made from the thread that currently handle all the reads from the socket. That will allow to ensure the `HeartBeatResponse` as the first thing sent back to ES on a `HeartBeatRequest`, and also better instrumentation / visibility into the ES exchanges.

**Comment by litch** [comment]
Btw, I have inside information that a big set of breaking changes to the  wire protocol is underway. 

I've also been able to get something built that uses the http client and long polling where the round trip message time is in single-digit milliseconds:  https://github.com/obsidian-btc/vertx-eventstore-http-polling

It uses a different http adapter that is vertx specific so is not very generalizable, but the long poll strategy is pretty great.  It winds up getting a lot of performance without all the complexity of the socket chatter. 

Sent from my iPhone

**Comment by mathieuravaux** [comment]
@litch I've been hit by this issue too, it's fixed as of this commit: 03243f9126cfcb312d78b42421b4e7fd2fd87c10

FWIW, I've been running this code in production for a few weeks now without any issue.

**Comment by litch** [comment]
Great to hear, thanks!
